### PR TITLE
Background image: size controls should show when an image is set

### PR DESCRIPTION
**Pull request by ramonjd**
Part of 

- https://github.com/WordPress/gutenberg/issues/54336

## What?
In the site editor, show the background size panel by default if an image value is present. 🎉 

Props to @andrewserong for [noticing](https://github.com/WordPress/gutenberg/pull/61271#issuecomment-2095236976) 

## Why?

So that the global styles control is consistent with other global styles. 

It should be, hence why this PR is labelled as "bug".

## How?

Checking for inherited background image value (from theme.json) as well as user-defined style value. If either is present, show the control unless background size is deactivated via theme.json settings.

## Testing Instructions

**TL;DR** In the site editor, for top-level global styles, the background size controls should show where there is an image value.

#### Background size controls are shown by default 

```json
{
	"$schema": "../../schemas/json/theme.json",
	"version": 2,
	"settings": {
		"appearanceTools": true
	},
	"styles": {
		"background": {
			"backgroundImage": {
				"url": "https://images.pexels.com/photos/22484288/pexels-photo-22484288/free-photo-of-the-circular-stone-terraces-of-the-inca-ruins.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2"
			}
		}
	}
}

```

#### Background size controls are displayed with theme.json value

```json
{
	"$schema": "../../schemas/json/theme.json",
	"version": 2,
	"settings": {
		"appearanceTools": true
	},
	"styles": {
		"background": {
			"backgroundImage": {
				"url": "https://images.pexels.com/photos/22484288/pexels-photo-22484288/free-photo-of-the-circular-stone-terraces-of-the-inca-ruins.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2"
			},
			"backgroundSize": "contain"
		}
	}
}
```

#### Background size controls are completely disabled

```json
{
	"$schema": "../../schemas/json/theme.json",
	"version": 2,
	"settings": {
		"appearanceTools": true,
		"background": {
			"backgroundSize": false
		}
	},
	"styles": {
		"background": {
			"backgroundImage": {
				"url": "https://images.pexels.com/photos/22484288/pexels-photo-22484288/free-photo-of-the-circular-stone-terraces-of-the-inca-ruins.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2"
			},
			"backgroundSize": "contain"
		}
	}
}
```

Also check that things work with your own site-wide background image, uploaded in global styles

Thank you for testing this PR. 🎖️ 

https://github.com/WordPress/gutenberg/assets/6458278/0375a676-b3a7-44b4-9aac-2a5cb1337746

### CODE REVIEW

File: packages/edit-site/src/components/global-styles/background-panel.js [hunk 1]
```diff
@@ -31,9 +31,7 @@ export default function BackgroundPanel() {
 
 	const defaultControls = {
 		backgroundImage: true,
-		backgroundSize:
-			!! style?.background?.backgroundImage &&
-			!! inheritedStyle?.background?.backgroundImage,
+		backgroundSize: false,
```
Comment (andrewserong) [inline packages/edit-site/src/components/global-styles/background-panel.js]: This means that the position controls aren't showing if a background image is set, with no size controls. Was the intention for them to be hidden? Apologies if I missed some of the evolution of the design discussion, but I thought the behaviour in global styles was intended that if you upload or add an image (or if an image is set via `theme.json`) that the size controls would always be visible.

That way when folks go to add an image, or to this part of the UI, the controls to adjust the background image are easily accessible, with the only time the controls are hidden being when there is no image.

**Review by andrewserong — COMMENTED** [review]
Thanks for following-up, but I think this might be looking to fix something different to what I had in mind in my review comment:

> One subtle thing I noticed in the UI is that it doesn't expose the size controls by default when the theme.json sets an image. Should it, do you think?

What I had in mind was that it seemed off to me when adding an image in global styles, and for the size controls not be shown by default after adding an image. Apologies if I missed the discussion where this was changed!

Edit: alternately I may have messed up my testing environment here, which is another possibility, as I'm running into a number of unrelated errors locally. I'll give this another test tomorrow 🙂

File: packages/edit-site/src/components/global-styles/background-panel.js (hunk 1 above)
Comment (ramonjd) [reply packages/edit-site/src/components/global-styles/background-panel.js]: >  Apologies if I missed some of the evolution of the design discussion, but I thought the behaviour in global styles was intended that if you upload or add an image (or if an image is set via theme.json) that the size controls would always be visible.

That was how things started out, but it was changed to be the same UX/UI as block supports.

See: https://github.com/WordPress/gutenberg/pull/60264#issuecomment-2032396220

**Comment by ramonjd** [comment]
> Edit: alternately I may have messed up my testing environment here, which is another possibility, as I'm running into a number of unrelated errors locally. I'll give this another test tomorrow

No, you're seeing things right. 

I was both assuming knowledge and reading your sentence wrong, sorry!! 

Since [this PR and this conversation](https://github.com/WordPress/gutenberg/pull/60264#issuecomment-2032396220), I've tried to keep the UX/UI the same between global styles and block supports. 

> One subtle thing I noticed in the UI is that it doesn't expose the size controls by default when the theme.json sets an image. Should it, do you think?

So I immediately jumped to a conclusion when you mentioned that the size controls weren't showing by default and my brain replaced "image" with "size". 😅 

If you think we should revert again to what you expected, we could open the design discussion up again?

**Comment by andrewserong** [comment]
> Since https://github.com/WordPress/gutenberg/pull/60264#issuecomment-2032396220, I've tried to keep the UX/UI the same between global styles and block supports.

Thanks for clarifying!

> If you think we should revert again to what you expected, we could open the design discussion up again?

Sorry to be a pain, but yes, I'd like to hear more from @jameskoster or @richtabor on this one. To me, having the size controls hidden by default feels inconsistent with other global styles controls. For spacing and typography, for example, at the block level in the post editor, we often keep things hidden by default, but in global styles they're all exposed by default.

When someone goes to add a site-wide background image, I think it's quite likely they'll want to tweak the image in some way, so I'd preference showing size controls by default once an image is added in global styles. Totally happy for us to go in a different direction if designers prefer, but I think it's worth double checking again, if that's okay 🙂

**Comment by ramonjd** [comment]
> To me, having the size controls hidden by default feels inconsistent with other global styles controls. For spacing and typography, for example, at the block level in the post editor, we often keep things hidden by default, but in global styles they're all exposed by default.

No worries. Whichever way we go, it's an easy change 👍🏻

**Comment by jameskoster** [comment]
> What I had in mind was that it seemed off to me when adding an image in global styles, and for the size controls not be shown by default after adding an image

Agree with this, the behavior on trunk for the 'site' background image in global styles feels right to me.

I know it's separate, but I still think putting the Size (and maybe other) settings in a popover is worth a try, like this old mockup:

Screenshot 2024-05-07 at 13 16 24

That would essentially solve this issue. Also; these settings currently take up way too much space in the Inspector given they're basically set-and-forget. I couldn't see if this was tracked in https://github.com/WordPress/gutenberg/issues/54336 at all.

**Comment by ramonjd** [comment]
> Agree with this, the behavior on trunk for the 'site' background image in global styles feels right to me.

Good stuff, thanks for confirming. I'll update this PR accordingly.

> Also; these settings currently take up way too much space in the Inspector given they're basically set-and-forget. I couldn't see if this was tracked in https://github.com/WordPress/gutenberg/issues/54336 at all.

Based on your designs, I have a WIP/experimental/explosive PR here, which is linked from the tracking issue (under General/not-yet-prioritized tasks). 

- https://github.com/WordPress/gutenberg/pull/60151

It won't be in WordPress 6.6 as it involves quite a bit of faffing about with ToolsPanel integration, and there are a bunch of unknowns (at least to me) in that area.

I plan to pick it up again and investigate further.

**Comment by ramonjd** [comment]
Okay folks, I've updated this PR to ensure global styles exposes the size controls by default when the theme.json sets an image 🤞🏻 

Thanks!

File: packages/block-editor/src/components/global-styles/background-panel.js [hunk 2]
```diff
@@ -379,7 +379,9 @@ function BackgroundSizeToolsPanelItem( {
 		( currentValueForToggle === 'cover' && repeatValue === undefined )
 	);
 
-	const hasValue = hasBackgroundSizeValue( style );
+	const hasValue =
+		hasBackgroundSizeValue( inheritedValue ) ||
+		hasBackgroundSizeValue( style );
```
Comment (ramonjd) [inline packages/block-editor/src/components/global-styles/background-panel.js]: Note to self:

There's an potential design edge case, where theme.json specifies a `background-image` string value for `background.backgroundImage`, e.g., `url('http://test.org/image.png')`, the controls won't display a preview because the value isn't being parsed.

I think this is correct as we can't possibly parse all possible values, e.g., linear gradients, image urls and multiples of both, but maybe we could pass a dummy image to the position controls at least, e.g., "No preview available" or something, or even disable the position drag controls.

Also, the copy could be changed from "Add background image" to "Update background image" when we detect a string value for `background.backgroundImage`, e.g., 

```js
	const previewLabel =
		title ||
		( ! style?.background?.backgroundImage?.url &&
		typeof inheritedValue?.background?.backgroundImage === 'string'
			? __( 'Update background image' )
			: title );
```

Comment (andrewserong) [inline packages/block-editor/src/components/global-styles/background-panel.js]: TL;DR, I think the original one liner of `const hasValue = hasBackgroundSizeValue( style );` might work better, but please let me know if I've missed why the change is necessary here! And now for the rambling thoughts while I was testing:

---

I think there might be a subtlety to the check here that isn't quite right yet. In testing, I noticed that if I set `"backgroundSize": "contain"`, then the UI of the tools panel menu indicates that a value is set when it shouldn't, i.e. the Reset button shouldn't be visible, but it is:

image

Then, if I hit Reset and go to make a change to the size values, the menu doesn't think any change occurred, and I can't reset it:

image

I was wondering if the logic for `hasValue` needs to be something more like:

```
if has a style value, and any of the values in the style value differs to the inherited value, then a value is set
```

I thought one example could (potentially) be something like:

```
	const hasValue =
		hasBackgroundSizeValue( style ) &&
		( style?.background?.backgroundSize !==
			inheritedValue?.background?.backgroundSize ||
			style?.background?.backgroundPosition !==
				inheritedValue?.background?.backgroundPosition ||
			style?.background?.backgroundRepeat !==
				inheritedValue?.background?.backgroundRepeat );
```

However, in testing that didn't quite work either. I tried hacking around, but I think there's an additional bit of complexity because the `inheritValue` appears to include any updates that have happened in the UI, and isn't the value of what's currently in `theme.json`.

So, after all that, I tried reinstating the simple one-liner of `const hasValue = hasBackgroundSizeValue( style );`, and in practice, that seemed to more closely resemble the expected behaviour for me. Most of the time it'll show the Reset button if a value is set, and not if one isn't set. If you happen to change the value back to one that overlaps with the value from `theme.json`, it still thinks that a change has been made, but if you press Reset, it gets back to normal.

What do you think? Please let me know if none of that makes sense, I think I wound up going around in circles a bit trying to wrap my head around it 😄

Comment (ramonjd) [reply packages/block-editor/src/components/global-styles/background-panel.js]: Thanks @andrewserong - I see some of what you're demonstrating. I've reverted the inherited style check for now - to be honest I think I left that in there by mistake during tests. Too many plates today. 🧇 

Seems to work a little more closely to what you'd expect.

I'll take a closer look at the rest of the feedback in a bit later. 

Cheers 🍺 !

Comment (andrewserong) [reply packages/block-editor/src/components/global-styles/background-panel.js]: Cheers, with that line reverted back to what it is in trunk, this is working very close to what I'd expect. I'd leave it as-is for now, and we can always revisit this in more detail further down the track. IMO this is close enough to perfect to be good enough for now 🙂

**Review by andrewserong — APPROVED** [review]
Thanks for all the back and forth here, this is testing great for me now!

LGTM 🎉

**Comment by ramonjd** [comment]
I appreciate your help @andrewserong and @jameskoster !
